flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/batch/sql/DataSetSingleRowJoinTest.scala

```diff
@@ -187,9 +187,155 @@ class SingleRowJoinTest extends TableTestBase {
       ),
       term("where", "AND(<(a1, b1)", "=(a2, b2))"),
       term("join", "a1", "a2", "b1", "b2"),
-      term("joinType", "NestedLoopJoin")
+      term("joinType", "NestedLoopInnerJoin")
     )
 
     util.verifySql(query, expected)
   }
+
+  @Test
+  def testSingleRowJoinLeftOuterJoin(): Unit = {
+    val util = batchTestUtil()
+    util.addTable[(Long, Int)]("A", 'a1, 'a2)
+    util.addTable[(Int, Int)]("B", 'b1, 'b2)
+
+    val queryLeftJoin =
+      "SELECT a2 FROM A " +
+        "LEFT JOIN " +
+        "(SELECT COUNT(*) AS cnt FROM B) " +
+        "AS x " +
+        "ON a1 = cnt"
+
+    val expected =
+      unaryNode(
+        "DataSetCalc",
+        unaryNode(
+          "DataSetSingleRowJoin",
+          batchTableNode(0),
+          term("where", "=(a1, cnt)"),
+          term("join", "a1", "a2", "cnt"),
+          term("joinType", "NestedLoopLeftJoin")
+        ),
+        term("select", "a2")
+      ) + "\n" +
+        unaryNode(
+          "DataSetAggregate",
+          unaryNode(
+            "DataSetUnion",
+            unaryNode(
+              "DataSetValues",
+              unaryNode(
+                "DataSetCalc",
+                batchTableNode(1),
+                term("select", "0 AS $f0")),
+              tuples(List(null)), term("values", "$f0")
+            ),
+            term("union", "$f0")
+          ),
+          term("select", "COUNT(*) AS cnt")
+        )
+
+    util.verifySql(queryLeftJoin, expected)
+  }
+
+  @Test
+  def testSingleRowJoinRightOuterJoin(): Unit = {
+    val util = batchTestUtil()
+    util.addTable[(Long, Int)]("A", 'a1, 'a2)
+    util.addTable[(Int, Int)]("B", 'b1, 'b2)
+
+    val queryRightJoin =
```

Comment:
Because the optimizer decides to execute the join as a regular join (see `DataSetJoin` in the plan). If we change the predicate to a non-equality predicate (`a1 < cnt`) the regular join cannot be used, and we can force a single row join.